tools/common/lib.go

```diff
@@ -1079,3 +1080,188 @@ func getIfSetString(c *cli.Context, p string, setCount *int) (r *string) {
+
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return client, nil
+}
+
+func (t *storeClientCache) close() {
+	for _, client := range t.cache {
+		client.Close()
+	}
+}
+
+// SealConsistencyCheck implement for show msg command line
+func SealConsistencyCheck(c *cli.Context, mClient mcli.Client) {
+
+	prefix := c.String("prefix")
+	seal := c.BoolT("seal")
+	verbose := c.BoolT("verbose")
+	veryVerbose := c.BoolT("veryverbose")
+
+	storeClients := newStoreClientCache(mClient)
+	defer storeClients.close()
+
+	const (
+		seqNumBits       = 26
+		seqNumBitmask    = (int64(1) << seqNumBits) - 1
+		timestampBitmask = math.MaxInt64 &^ seqNumBitmask
+		sealKeyTimestamp = math.MaxInt64 & timestampBitmask
+	)
+
+	req := &shared.ListDestinationsRequest{
+		Prefix: common.StringPtr(prefix),
+		Limit:  common.Int64Ptr(DefaultPageSize),
+	}
+
+	for {
+		resp, err := mClient.ListDestinations(req)
+		ExitIfError(err)
+
+		for _, desc := range resp.GetDestinations() {
+
+			destUUID := desc.GetDestinationUUID()
+
+			// fmt.Printf("destUUID=%v\n", destUUID)
+
+			listExtentsStats := &shared.ListExtentsStatsRequest{
+				DestinationUUID: common.StringPtr(string(destUUID)),
+				Limit:           common.Int64Ptr(DefaultPageSize),
+			}
+
+			for {
+				listExtentStatsResult, err1 := mClient.ListExtentsStats(listExtentsStats)
+				ExitIfError(err1)
+
+				for _, stats := range listExtentStatsResult.ExtentStatsList {
+
+					extent := stats.GetExtent()
+					extentUUID := extent.GetExtentUUID()
+
+					if stats.GetStatus() == shared.ExtentStatus_SEALED {
+
+						storeUUIDs := extent.GetStoreUUIDs()
+
+						for _, storeUUID := range storeUUIDs {
+
+							storeClient, err1 := storeClients.get(storeUUID)
+
+							if err1 != nil {
+								fmt.Printf("store=%v: error getting store client: %v\n", storeUUID, err1)
+								continue
+							}
+
+							req := store.NewGetAddressFromTimestampRequest()
+							req.ExtentUUID = common.StringPtr(string(extentUUID))
+							req.Timestamp = common.Int64Ptr(sealKeyTimestamp)
+
+							// query storage to find address of the message with the given timestamp
+							resp, err1 := storeClient.GetAddressFromTimestamp(req)
+
+							var extentNotFound bool
+
+							if err1 != nil {
+								_, extentNotFound = err1.(*store.ExtentNotFoundError)
+
+								if !extentNotFound {
+									fmt.Printf("dest=%v extent=%v store=%v: GetAddressFromTimestamp error=%v\n",
+										destUUID, extentUUID, storeUUID, err1)
+									continue
+								}
+							}
+
+							switch {
+							case extentNotFound || !resp.GetSealed(): // handle un-sealed extent
+
+								fmt.Printf("dest=%v extent=%v store=%v: ", destUUID, extentUUID, storeUUID)
+
+								if extentNotFound {
+									fmt.Printf("not sealed (extent missing)\n", destUUID, extentUUID, storeUUID)
+								} else {
+									fmt.Printf("not sealed\n", destUUID, extentUUID, storeUUID)
+								}
+
+								// now seal the extent
+								if seal {
```

Comment:
yes; there are scenarios where a store does not even know about an extent .. but it has already been sealed. sending a seal request to a store for such an extent, creates that extent and seals it.